### ISSUE TITLE
How to output result image from output?

### ISSUE THREAD

**Issue by OOXXXXOO** [jfzhang95/pytorch-deeplab-xception#28]
When i finished the model training. & load  model & input a single image
by:
```python
# Image preprocessing
    image = cv2.imread(image_path, cv2.IMREAD_COLOR).astype(float)
    image = cv2.resize(image, dsize=(512,512))
    image_original = image.astype(np.uint8)
    image = torch.from_numpy(image.transpose(2, 0, 1)).float().unsqueeze(0)
    image = image.to(device)

    # Inference
    output = Model(image)
```
output is a tensor & shape is (5,512,512).
like:
```python
[[[  4916.116    5713.0825   6510.0493 ...   8198.43     7449.5635
     6700.7207]
  [  5698.8423   6637.15     7575.4585 ...   9693.137    8787.99
     7882.872 ]
  [  6481.569    7561.218    8640.867  ...  11187.843   10126.417
     9065.023 ]
  ...
```
# if i want to invert the tensor to labelimage ,What should i do? 
if i process output by **softmax or argmax or torch.max** ,the output will be null image like
```python
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
```
It's a weird question!

**Comment by jfzhang95** [comment]
What is your model setting? 
When I run: 
```
    model = DeepLab(backbone='resnet', output_stride=16, num_classes=21).cuda()
    model.eval()
    input = torch.rand(1, 3, 513, 513).cuda()
    output = model(input)
    print(output.size())
```
The output size is (1, 21, 513, 513).
Then run the following code
```
    prediction = output.data.cpu().numpy()
    prediction = np.argmax(prediction, axis=1)
```
I could get the final prediction and the shape of prediction is (513, 513).

Based on the example you provided, I guess the number of classes in your example is 5. So, when you get output, you could run ```pred=output.data.cpu().numpy()``` and ```pred=np.argmax(pred, axis=0)``` to get the prediction labels, just like what I do above.

Hope this could help you :)

**Comment by OOXXXXOO** [comment]
Thank you！But：
Of course ，My default num_class set to 5，So The output tensor size like（1,5.512,512）& output is valid data
When i add 
```python
    prediction = output.data.cpu().numpy()
    prediction = np.argmax(prediction, axis=1)
```
the prediction also will be：
```python
[[[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]]
```
When i add `prediction.sum()` it's will equal to 0
It alway is a  null ndarray.................
In addition, if i open my tensorboard, all the prediction is valid data like
 
![image](https://user-images.githubusercontent.com/11496629/48999909-edaf2d80-f192-11e8-859b-456df40aac38.png)
So the model file is wrok, I just couldn't invert the output tensor to prediction image by any method .(include softmax ,argmax,torch.max)
Maybe  you have some solution  ?

**Comment by jfzhang95** [comment]
I used PIL to read image, so the image mode is RGB instead of BGR, if you use opencv, maybe you should run ```img = img[:,:,::-1]``` to transfer image from BGR to RGB.

**Comment by Venka97** [comment]
Is it necessary to normalize the image for prediction the mask? eg. If I have an image - img1, do I have to divide it by 255 and subtract the mean and std dev from it before inputting it into the model?

**Comment by albertchristianto** [comment]
hi everyone,

right now i am writing inference code for this repo, but i face same problem that @OOXXXXOO had faced. Have you solve the problem? how did you do to solve it?

thank you very much

best regards,
albert christianto

**Comment by leedoge** [comment]
> hi everyone,
> 
> right now i am writing inference code for this repo, but i face same problem that @OOXXXXOO had faced. Have you solve the problem? how did you do to solve it?
> 
> thank you very much
> 
> best regards,
> albert christianto

hello!i also have faced some problem,have you finished your inference code?

**Comment by albertchristianto** [comment]
@leedoge hi,
I think i have solved this problem. but, maybe no.  i forgot because it has already 1 month.
You can check your input processing. 
like have you substract your image with imagenet mean ?

best regards,
Albert Christianto
